### PR TITLE
Enroll node API 

### PR DESCRIPTION
**Pull request by jkakavas**
Enroll node API can be used by new nodes in order to join an
existing cluster that has security features enabled. The response
of a call to this API contains all the necessary information that
the new node requires in order to configure itself and bootstrap
trust with the existing cluster.

### CODE REVIEW

File: docs/reference/cluster/enroll-node.asciidoc [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+==== {api-request-title}
+
+`POST /_cluster/enroll_node`
+
+`PUT /_cluster/enroll_node`
+
+[[cluster-enroll-node-api-prereqs]]
+==== {api-prereq-title}
+
+* You must have the `enroll` <<privileges-list-cluster,cluster privilege>> to use this API.
```
Comment (jkakavas) [inline docs/reference/cluster/enroll-node.asciidoc:19]: Privilege is introduced in a followup PR

File: docs/reference/cluster/enroll-node.asciidoc [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+++++
+<titleabbrev>Enroll Node</titleabbrev>
+++++
+
+Allows a new node to join an existing cluster with security enabled.
+
+[[cluster-enroll-node-api-request]]
+==== {api-request-title}
+
+`GET /_cluster/enroll_node`
```
Comment (BigPandaToo) [inline docs/reference/cluster/enroll-node.asciidoc:12]: Isn't it POST?

File: docs/reference/cluster.asciidoc [hunk 3]
```diff
@@ -78,6 +78,8 @@ GET /_nodes/ra*:2*
 
 include::cluster/allocation-explain.asciidoc[]
 
+include::cluster/enroll-node.asciidoc.asciidoc[]
```
Comment (BigPandaToo) [inline docs/reference/cluster.asciidoc:81]: extra.asciidoc

**Comment by BigPandaToo** [comment]
Just added couple comments, but will take a closer look on Monday

File: docs/reference/cluster/enroll-node.asciidoc (hunk 2 above)
Comment (jkakavas) [reply docs/reference/cluster/enroll-node.asciidoc:12]: Nope, it is a GET. It doesn't change anything in the node, it just retrieves information. It might even make sense to rename this to Get Enrollment Information or something similar

**Comment by jkakavas** [comment]
HLRC changes incoming

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/XPackPlugin.java [hunk 4]
```diff
@@ -331,7 +334,8 @@ public Settings additionalSettings() {
         handlers.add(new RestXPackUsageAction());
         handlers.add(new RestReloadAnalyzersAction());
         handlers.addAll(licensing.getRestHandlers(settings, restController, clusterSettings, indexScopedSettings, settingsFilter,
-                indexNameExpressionResolver, nodesInCluster));
+            indexNameExpressionResolver, nodesInCluster));
+        handlers.add(new RestNodeEnrollmentAction());
```
Comment (ywangd) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/XPackPlugin.java:338]: This rest handler and the corresponding transport handler are registered here in xpack core instead of security. But I wonder whether it would be better to register them under xpack security because this API assumes security being enabled (which really is by design).

When we bootstrap a new cluster, the init process manipulates the `elasticsearch.yml` file. So we have the guarantee that security is enabled. But we don't have the same guarantee for an existing cluster. IIUC, users can still explicitly disable security in v8.0. So for a cluster where security is explicitly disabled, it does not seem to have much sense to keep this API around? It's likely error out because there may not be any keystore file for it to read. It would probably be even worse if there are keystore files (maybe leftovers before security is disabled) because a successful response could give false impression of security. 

If we register this API under security, it will not be available when security is explicitly disabled which naturally solves the aforementioned issue. One advantage that I can think of for having it under core is that we can return an error response says "please enable security". But this can probably be replaced by documentation and also we don't have this type of error messages for any other security APIs.

In summary, I'd suggest to register this API and likely other related APIs, as well as the setting for enrollment mode, under security.

Comment (jkakavas) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/XPackPlugin.java:338]: I went back and forth on this decision but settled on core as these APIs felt more like cluster APIs than strictly security APIs. With that in mind it made more sense to have these in core. 

> When we bootstrap a new cluster, the init process manipulates the elasticsearch.yml file. So we have the guarantee that security is enabled. But we don't have the same guarantee for an existing cluster. IIUC, users can still explicitly disable security in v8.0. So for a cluster where security is explicitly disabled, it does not seem to have much sense to keep this API around?  

Security will default to being enabled so the init process changes to the elasticsearch.yml make no difference here. But you are right, users are still allowed to disable security if it fits their use case. We will be introducing an EnrollmentBaseRestHandler of sorts for all the Rest actions where we'd be checking if security is enabled and if enrollment mode is enabled so this solves the issue.

I don't have a strong argument in favor of keeping this in core, but I don't see a strong argument in favor of moving these to the security plugin either. Happy to discuss further though

Comment (ywangd) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/XPackPlugin.java:338]: Sorry if this has already been a deliberated decision. I wasn't aware of the context. I am happy as long as it is an intentional change. Thanks!
> We will be introducing an EnrollmentBaseRestHandler of sorts for all the Rest actions where we'd be checking if security is enabled and if enrollment mode is enabled so this solves the issue.

If providing a more useful error message (more useful than "handler not found") is the goal, having it registered in core makes sense. I still do think this API is closer to security than core mainly by the fact that it does not work unless security is enabled. Though it's a rather weak argument for this API. What about the other API where an API key needs to be created (#72186)? Do we want to have all APIs located in the same package, i.e. core? It's possible to coordinate the API key creation in core as well, i.e rely only on the CreateApiKey Request and Response classes. So far, we do not seem to have any code creating API key outside of the security package. This might be the chance to set up an example. With our current stance towards security, I guess the ideal world is to move basic security features into `core`, which could help get rid of these discussion for good and make it easier if another plugin wants to leverage it (maybe worth to discuss separately). 

> Security will default to being enabled so the init process changes to the elasticsearch.yml make no difference here.

It maybe unlikely, but theoretically still possible that security is explicity disabled in `elasticsearch.yml` before the init process runs (maybe the user was trying to experiment before starting the init process). For the process to be foolproof, init might want to check and correct the setting if needed.

Comment (jkakavas) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/XPackPlugin.java:338]: > Sorry if this has already been a deliberated decision. I wasn't aware of the context. I am happy as long as it is an intentional change. Thanks!

No worries, always happy to revisit decisions

> If providing a more useful error message (more useful than "handler not found") is the goal, 

It's a nice effect of it, but not a hard requirement or a goal I guess. 

> It's possible to coordinate the API key creation in core as well, i.e rely only on the CreateApiKey Request and Response classes

Right! That's what we're doing so far. 

> It maybe unlikely, but theoretically still possible that security is explicity disabled in elasticsearch.yml before the init process runs (maybe the user was trying to experiment before starting the init process). For the process to be foolproof, init might want to check and correct the setting if needed.

Yes absolutely. My point was that the init process makes no difference. Whether it runs or not, the user can set the YAML file settings as they please, before or after it runs, so I agree with you that there is a chance that security is disabled and we should handle this gracefully.

Comment (jkakavas) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/XPackPlugin.java:338]: Thanks again for starting this discussion @ywangd ! I chatted with Tim this morning about this and came to the conclusion that we should move these APIs to the security plugin because; 

- In genera, moving forward we want to put things in plugins instead of core
- These APIs cannot be used when security is off ( Create enrollment token generates an API key and enroll client and enroll node validate one ) 
- We can select to have a user friendly way of notifying the users that the APIs cannot be used because security is off ( rather than returning a 404 ), and this can be a more generic approach for the rest of the security APIs too. We could register the RestHandlers even when security is off and return a helpful error message there.

**Comment by jkakavas** [comment]
@elasticmachine update branch

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/enrollment/TransportNodeEnrollmentAction.java [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+                nodeList.add(node.getAddress().toString());
+            }
+            listener.onResponse(new NodeEnrollmentResponse(httpCaKey,
+                httpCaCert,
+                transportKey,
+                transportCert,
+                clusterService.getClusterName().value(),
+                nodeList));
+
+        } catch (Exception e) {
```
Comment (albertzaharovits) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/enrollment/TransportNodeEnrollmentAction.java:104]: Nit: this is probably one of the few places where you don't have to worry about throwing exceptions if the method takes an action listener.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ssl/StoreKeyConfig.java [hunk 6]
```diff
@@ -154,6 +155,26 @@ X509ExtendedTrustManager createTrustManager(@Nullable Environment environment) {
         }
     }
 
+    public List<Tuple<PrivateKey, X509Certificate>> getPrivateKeyEntries(Environment environment) {
```
Comment (albertzaharovits) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ssl/StoreKeyConfig.java:158]: Is this method necessary (as well as making the class public, and using `instanceof` in the transport handler)?
I've found https://github.com/elastic/elasticsearch/blob/0220dfb3fe4de4ca416918a76a6fdcff0369f624/libs/ssl-config/src/test/java/org/elasticsearch/common/ssl/StoreKeyConfigTests.java#L138 that appears to do the same thing. Have you looked at that?

File: client/rest-high-level/src/main/java/org/elasticsearch/client/SecurityClient.java [hunk 7]
```diff
@@ -1130,4 +1132,27 @@ public Cancellable delegatePkiAuthenticationAsync(DelegatePkiAuthenticationReque
         return restHighLevelClient.performRequestAsyncAndParseEntity(request, SecurityRequestConverters::delegatePkiAuthentication, options,
                 DelegatePkiAuthenticationResponse::fromXContent, listener, emptySet());
     }
+
+
+    /**
+     * Allows a node to join to a secured cluster using the Enroll Node API.
```
Comment (albertzaharovits) [inline client/rest-high-level/src/main/java/org/elasticsearch/client/SecurityClient.java:1138]: nit: `secured` -> `secure`

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/operator/OperatorOnlyRegistry.java [hunk 8]
```diff
@@ -35,7 +36,8 @@
         // Repository analysis actions are not mentioned in core, literal strings are needed.
         "cluster:admin/repository/analyze",
         "cluster:admin/repository/analyze/blob",
-        "cluster:admin/repository/analyze/blob/read"
+        "cluster:admin/repository/analyze/blob/read",
+        NodeEnrollmentAction.NAME
```
Comment (albertzaharovits) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/operator/OperatorOnlyRegistry.java:40]: Does this mean that invoking client-soon-to-be-node has to be an operator?
If it's using an API key it is not currently possible. Even if it where, it doesn't sound to useful. When we release the enroll credential we explicitly want any client with that credential to be able to use it to join the cluster.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/enrollment/NodeEnrollmentAction.java [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.core.security.action.enrollment;
+
+import org.elasticsearch.action.ActionType;
+
+public final class NodeEnrollmentAction extends ActionType<NodeEnrollmentResponse> {
+
+    public static final String NAME = "cluster:admin/xpack/enroll";
```
Comment (albertzaharovits) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/enrollment/NodeEnrollmentAction.java:14]: Barring the enroll credential, no one else should be able to invoke this API (because it is not useful and because you get the HTTP CA key and can impersonate the cluster).
I suggest we exclude this action from all the existing privileges; so that it has to be granted explicitly.

File: client/rest-high-level/src/main/java/org/elasticsearch/client/SecurityClient.java (hunk 7 above)
Comment (albertzaharovits) [inline client/rest-high-level/src/main/java/org/elasticsearch/client/SecurityClient.java:1138]: I would highlight that this API allows the caller to stand-up a node that is able to join the secure cluster, which also means that the caller has the capacity to "impersonate" the cluster to any other clients.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/enrollment/TransportNodeEnrollmentAction.java [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+            }
+
+            final String httpCaKey = Base64.getUrlEncoder().encodeToString(httpCaKeysAndCertificates.get(0).v1().getEncoded());
+            final String httpCaCert = Base64.getUrlEncoder().encodeToString(httpCaKeysAndCertificates.get(0).v2().getEncoded());
+            final String transportKey = Base64.getUrlEncoder().encodeToString(transportKeysAndCertificates.get(0).v1().getEncoded());
+            final String transportCert = Base64.getUrlEncoder().encodeToString(transportKeysAndCertificates.get(0).v2().getEncoded());
+            final List<String> nodeList = new ArrayList<>();
+            for (DiscoveryNode node : clusterService.state().getNodes()) {
+                nodeList.add(node.getAddress().toString());
+            }
```
Comment (albertzaharovits) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/enrollment/TransportNodeEnrollmentAction.java:96]: What do you think if we use the output of `NodeEnrollmentAction.NAME` with the transport metric parameter?
It gives me more confidence that we're returning the right addresses, and the client might already be used to parsing that.

**Review by albertzaharovits — COMMENTED** [review]
I had a couple of comments, but overall it does what it's supposed to do.
We can discuss it a bit, but we might need to polish it when all the pieces are in-place, so lets not overdo it now.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/enrollment/TransportNodeEnrollmentAction.java (hunk 5 above)
Comment (jkakavas) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/enrollment/TransportNodeEnrollmentAction.java:104]: Can you elaborate ? The idea here was to catch any Exceptions that might get thrown so that we call the action listener. On second thought though now, it's only the `getEncoded()` that would throw in the code above so I can have more contained try,catch and return a useful error with the exception

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ssl/StoreKeyConfig.java (hunk 6 above)
Comment (jkakavas) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ssl/StoreKeyConfig.java:158]: The instanceof is there so that we know that this key manager was built from a keystore and not from PEM files. The idea was that the enrollment process depends on our self-generated key material and is not supposed to be an one-solution-fits-all thing, and this was an additional check to that effect. 

It is not strictly needed though, I could remove this if making StoreKeyConfig public is something that you find bothersome. 

> that appears to do the same thing. Have you looked at that?

I don't think you can iterate all the PrivateKeyEntries from the KeyManager, we'd need to know the alias name. Which we can know because we will be generating the keystores, and if we only support these, then it's doable. I'll take a look, thanks

File: client/rest-high-level/src/main/java/org/elasticsearch/client/SecurityClient.java (hunk 7 above)
Comment (jkakavas) [reply client/rest-high-level/src/main/java/org/elasticsearch/client/SecurityClient.java:1138]: It is `secured` on purpose. I wanted to say a cluster with security features enabled, not claim that such a cluster is "secure" . Maybe I'll just say "a cluster with security features enabled" :)

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/operator/OperatorOnlyRegistry.java (hunk 8 above)
Comment (jkakavas) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/operator/OperatorOnlyRegistry.java:40]: The idea is that this should not be used in ESS, even if someone is superuser or grants themselves the enroll privilege. We don't "support" operator privileges in any other circumstance, so I don't see this as a problem. 

> If it's using an API key it is not currently possible. 

I've missed that part, will take a look

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/enrollment/NodeEnrollmentAction.java (hunk 9 above)
Comment (jkakavas) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/enrollment/NodeEnrollmentAction.java:14]: yes! Thanks for calling me out on this. We have discussed this extensively and we don't want any other privilege to imply enroll / manage enrollment  or to give access to these actions. I'll move this over to a specific prefix, maybe simply `enrollment:enroll` for this

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/enrollment/TransportNodeEnrollmentAction.java (hunk 10 above)
Comment (jkakavas) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/enrollment/TransportNodeEnrollmentAction.java:96]: I'm not sure I understood the suggestion @albertzaharovits .You mean that we should return `enroll:enroll_node:192.168.1.2:9300` instead of `192.168.1.2:9300` ? What does this buy us ?

File: client/rest-high-level/src/main/java/org/elasticsearch/client/SecurityClient.java (hunk 7 above)
Comment (jkakavas) [reply client/rest-high-level/src/main/java/org/elasticsearch/client/SecurityClient.java:1138]: I'll add something to that effect. I think the most important part is to clearly call this out in the privilege description but no objection to adding information in the APIs description

Comment (jkakavas) [reply client/rest-high-level/src/main/java/org/elasticsearch/client/SecurityClient.java:1138]: @albertzaharovits can you elaborate on "which also means that the caller has the capacity to "impersonate" the cluster to any other clients. " ? What specific threat do you have in mind ?

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/operator/OperatorOnlyRegistry.java (hunk 8 above)
Comment (jkakavas) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/operator/OperatorOnlyRegistry.java:40]: Chatted a little with Yang. Wasn't aware of the API key limitation for operator only. In retrospect, we don't actually need the API keys to be operator only and since the enrollment mode settings is not going to be a dynamic one, then it can't also be operator only. I think we'd need to depend on setting the enrollment mode to false on ESS instances and not whitelist the setting so that administrators cannot overwrite it in ESS deployment settings

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/enrollment/NodeEnrollmentAction.java (hunk 9 above)
Comment (jkakavas) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/enrollment/NodeEnrollmentAction.java:14]: @albertzaharovits how would you feel about putting this under `cluster:admin/xpack/security/enroll` and restricting this from `manage_security`, vs leaving this as is and excluding it from the `manage` cluster ? Or should we go all the way and exclude this from `all` too ?

**Review by BigPandaToo — APPROVED** [review]
Approoving considering you address Albert's comments

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ssl/StoreKeyConfig.java (hunk 6 above)
Comment (jkakavas) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ssl/StoreKeyConfig.java:158]: I'll leave this as is @albertzaharovits as StoreConfig is moving to common/ssl only and it will be public there so I don't think there is any concern adding a helper method that does exactly what we want it to there. Let me know if you feel strongly otherwise

**Comment by jkakavas** [comment]
Thanks for the round @albertzaharovits . I addressed your feedback or commented accordingly, if we could do another round on Monday before I go, that'd be awesome <3

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/enrollment/TransportNodeEnrollmentAction.java (hunk 5 above)
Comment (albertzaharovits) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/enrollment/TransportNodeEnrollmentAction.java:104]: I understand the intention. I meant to say that you don't have to do it because it's already handled this way in https://github.com/elastic/elasticsearch/blob/668a72dec5dc06a56ba5753e093c1731143d5941/server/src/main/java/org/elasticsearch/action/support/TransportAction.java#L79 

In general, I'm not a fan of expansive try/catch blocks as the intention behind the catch is lost, it feels more like "ignore exceptions".

This is just a nit.

Comment (jkakavas) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/enrollment/TransportNodeEnrollmentAction.java:104]: Thanks Albert, makes sense. I'll limit the try block scope and have any unexpected ones handled by the base class.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/enrollment/NodeEnrollmentAction.java (hunk 9 above)
Comment (albertzaharovits) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/enrollment/NodeEnrollmentAction.java]: Please create an issue for this, since I understand this will be a follow-up.

File: client/rest-high-level/src/main/java/org/elasticsearch/client/SecurityClient.java (hunk 7 above)
Comment (albertzaharovits) [reply client/rest-high-level/src/main/java/org/elasticsearch/client/SecurityClient.java]: An astute user might infer that this endpoint is very sensitive, given that a rogue node can siphon all the data from the cluster (even this, I think, is worth mentioning somewhere).
But it's more than this, because the caller doesn't have to join the cluster (which might be otherwise monitored, and the attacker be detected), it can use the API response to forge himself certificates that can impersonate any cluster node to a client.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/enrollment/TransportNodeEnrollmentAction.java (hunk 10 above)
Comment (albertzaharovits) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/enrollment/TransportNodeEnrollmentAction.java:96]: Bleah... sorry, I messed up the clipboard buffers.
I meant `NodeEnrollmentAction.NAME` -> `NodesInfoAction.NAME` .

**Review by albertzaharovits — APPROVED** [review]

File: client/rest-high-level/src/main/java/org/elasticsearch/client/SecurityClient.java (hunk 7 above)
Comment (jkakavas) [reply client/rest-high-level/src/main/java/org/elasticsearch/client/SecurityClient.java]: @albertzaharovits worth noting that an attacker needs also to be in a position to MITM legitimate clients in order to impersonate the server ( assuming that by "impersonate any node" you mean that the caller of the API can get a valid signed certificate signed by the same CA that would have signed the legitimate nodes certificates )

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/enrollment/TransportNodeEnrollmentAction.java (hunk 10 above)
Comment (jkakavas) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/enrollment/TransportNodeEnrollmentAction.java:96]: makes sense, I'm calling this for the enroll client API transport action, might as well do this here too, no objections

**Comment by jkakavas** [comment]
Thanks for the round @albertzaharovits , I'll merge this if CI is green before I go, otherwise please do merge at will

File: client/rest-high-level/src/main/java/org/elasticsearch/client/SecurityClient.java (hunk 7 above)
Comment (albertzaharovits) [reply client/rest-high-level/src/main/java/org/elasticsearch/client/SecurityClient.java]: @jkakavas One must pass an enroll token to clients to have them configured; the same enroll token that's used to register nodes. There can be an attacker that intercepts client-cluster communication.

Comment (jkakavas) [reply client/rest-high-level/src/main/java/org/elasticsearch/client/SecurityClient.java]: Yes that is my argument too, one needs to be MITM. I'm not trying to downplay the attack, just noting the constraints
